pelita/layout.py

```diff
@@ -98,211 +103,46 @@ def get_layout_by_name(layout_name):
-
-
-    A layout string is composed of wall characters '#', food characters '.', and
-    bot characters '0', '1', '2', and '3'.
+def parse_layout(layout_str, food=None, bots=None):
+    """Parse a layout string.
 
-    Valid layouts must be enclosed by walls and be of rectangular shape. Example:
+    A valid layout string is enclosed by walls and rectangular:
 
      ########
-     #0  .  #
-     #2    1#
-     #  .  3#
+     #a  .  #
+     #b ## x#
+     #  .  y#
      ########
 
 
-    If items are overlapping, several layout strings can be concatenated:
-     ########
-     #0  .  #
-     #     1#
-     #  .  3#
-     ########
-     ########
-     #2  .  #
-     #     1#
-     #  .  3#
-     ########
-
-    In this case, bot '0' and bot '2' are on top of each other at position (1,1)
+    Return a dict
+        {'walls': list_of_wall_coordinates,
+         'food' : list_of_food_coordinates,
+         'bots'  : list_of_bot_coordinates in the order (a,x,b,y) }
 
-    If `allow_enemy_chars` is True, we additionally allow for the definition of
-    at most 2 enemy characters with the letters "E" and "?". The returned dict will
-    then additionally contain an entry "enemy" which contains these coordinates and
-    an entry "is_noisy" that specifies which of the given enemies is noisy.
-    If only one enemy character is given, both will be assumed sitting on the
-    same spot. """
+    In the example above:
+    {'walls': [(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (1, 0), (1, 4), (2, 0),
+               (2, 4), (3, 0), (3, 2), (3, 4), (4, 0), (4, 2), (4, 4), (5, 0),
+               (5, 4), (6, 0), (6, 4), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4)],
+     'food': [(3, 3), (4, 1)],
+     'bots': [(1, 1), (6, 2), (1, 2), (6, 3)]}
 
-    if allow_enemy_chars:
-        num_bots = 2
-    else:
-        num_bots = 4
+    Additional food and bots can be passed:
 
-    layout_list = []
-    start = False
-    for i, line in enumerate(layout_str.splitlines()):
-        row = line.strip()
-        if not row:
-            # ignore emptylines
-            continue
-        if not start:
-            # start a new layout
-            # check that row is a valid opening string
-            if row.count('#') != len(row):
-                raise ValueError(f"Layout does not start with a row of walls (line: {i})!")
-            current_layout = [row]
-            start = True
-            continue
-        # we are in the middle of a layout, just append to the current
-        # layout unless we detect the closing string
-        current_layout.append(row)
-        if row.count('#') == len(row):
-            # this is a closing string
-            # append the layout to the layout list
-            layout_list.append('\n'.join(current_layout))
-            start = False
+      - food: a list of coordinates of additional food pellets
+      - bots: a dictionary { char : (coord_x, coord_y)}, where char in 'a', 'b', 'x', 'y'
+    """
```

Comment:
Can you write a test for it? If a bot is specified in the string and in the dictionary, the coordinates in the dictionary win. I am adding a line to the doc-string to make it clear.

Reply:
I think it should raise a ValueError when it does not match. Otherwise the trick you mentioned earlier (to ensure that a bot sits on a food) doesn’t work.

Reply:
Postponed